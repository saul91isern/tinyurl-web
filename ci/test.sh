#!/bin/sh

echo "----- Install dependencies -----"
yarn
echo "----- Linter -----"
yarn test --ci

