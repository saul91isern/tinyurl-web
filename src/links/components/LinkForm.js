import _ from "lodash/fp";
import React from "react";
import { Button, Form } from "semantic-ui-react";
import { useForm, Controller } from "react-hook-form";
import isURL from "validator/lib/isURL";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { createLink } from "../routines";

export const LinkForm = ({ createLink }) => {
  const { handleSubmit, control, formState } = useForm({
    mode: "all",
    defaultValues: {
      url: "",
    },
  });
  const onSubmit = (value) => createLink(value);
  const { errors, isDirty } = formState;

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <Controller
        control={control}
        name="url"
        rules={{
          required: "Field url is required",
          maxLength: {
            value: 2048,
            message:
              "The url is too big. Maximum supported length of 2048 characters",
          },
          validate: (value) => isURL(value) || "Invalid url format",
        }}
        render={({ field: { onChange, value } }) => (
          <Form.Input
            id="url"
            fluid
            autoComplete="off"
            required
            label={"Url"}
            placeholder={"Enter a valid url"}
            onChange={(_e, { value }) => onChange(value)}
            error={
              isDirty &&
              !_.isNil(errors?.url?.message) && {
                content: errors.url.message,
                pointing: "below",
              }
            }
            value={value || ""}
          />
        )}
      />
      <Button type="submit" disabled={!_.isEmpty(errors) || !isDirty}>
        Submit
      </Button>
    </Form>
  );
};

LinkForm.propTypes = {
  createLink: PropTypes.func,
};

const mapDispatchToProps = { createLink };

export default connect(null, mapDispatchToProps)(LinkForm);
