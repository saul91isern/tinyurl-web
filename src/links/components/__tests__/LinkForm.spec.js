/* eslint-disable quotes */
import React from "react";
import { waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { render } from "../../../test/render";
import { LinkForm } from "../LinkForm";

describe("<LinkForm />", () => {
  const createLink = jest.fn();
  it("matches last snapshot", async () => {
    const { container } = render(<LinkForm createLink={createLink} />, {});
    expect(container).toMatchSnapshot();
  });

  it("creates link on form submit", async () => {
    const { findByPlaceholderText, queryByText } = render(
      <LinkForm createLink={createLink} />,
      {}
    );
    expect(queryByText(/Submit/)).toBeTruthy();
    expect(queryByText(/Submit/)).toBeDisabled();

    userEvent.type(
      await findByPlaceholderText(/Enter a valid url/),
      "https://foo.com/bar"
    );

    await waitFor(() => expect(queryByText(/Submit/)).toBeEnabled());

    userEvent.click(await queryByText(/Submit/));

    await waitFor(() =>
      expect(createLink).toHaveBeenCalledWith({ url: "https://foo.com/bar" })
    );
  });
});
