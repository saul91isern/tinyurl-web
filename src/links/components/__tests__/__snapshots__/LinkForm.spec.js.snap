// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkForm /> matches last snapshot 1`] = `
<div>
  <form
    class="ui form"
  >
    <div
      class="required field"
    >
      <label
        for="url"
      >
        Url
      </label>
      <div
        class="ui fluid input"
      >
        <input
          autocomplete="off"
          id="url"
          placeholder="Enter a valid url"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="ui disabled button"
      disabled=""
      tabindex="-1"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
